cloud_detect/providers/aws_provider.py

```diff
@@ -28,22 +32,34 @@ async def identify(self):
         self.logger.info('Try to identify AWS')
         return self.check_vendor_file() or await self.check_metadata_server()
 
+    async def _get_metadata_v2(self):
+        async with aiohttp.ClientSession() as session:
+            async with session.put(self.metadata_token_url, headers={'X-aws-ec2-metadata-token-ttl-seconds': '21600'}) as response:
+                token = await response.text()
+        return await self._get_metadata(headers={'X-aws-ec2-metadata-token': token})
+
+    async def _get_metadata(self, headers=None):
+        async with aiohttp.ClientSession() as session:
+            async with session.get(self.metadata_url, headers=headers) as response:
+                response.raise_for_status()
+                response = await response.json(content_type=None)
+                if response['imageId'].startswith('ami-', ) and response[
+                    'instanceId'
+                ].startswith('i-'):
+                    return True
+        return False
+
     async def check_metadata_server(self):
         """
             Tries to identify AWS via metadata server
         """
         self.logger.debug('Checking AWS metadata')
-        try:
-            async with aiohttp.ClientSession() as session:
-                async with session.get(self.metadata_url) as response:
-                    response = await response.json(content_type=None)
-                    if response['imageId'].startswith('ami-',) and response[
-                        'instanceId'
-                    ].startswith('i-'):
-                        return True
-            return False
-        except BaseException:
-            return False
+        results = await asyncio.gather(
+            self._get_metadata(),
+            self._get_metadata_v2(),
+            return_exceptions=True
```

Comment:
I added the `return_exceptions` because I think that it will make the code look cleaner, but I will add the try block instead if you prefer.
The `raise_for_status` is needed since the API call can return with 401 status and an empty body, I assume that `response.json` will also fail in this case 😃

Reply:
Another option is to replace the try block with `contextlib.suppress(BaseException)`
BTW, any reason you are using `BaseException` and not `Exception`?

Reply:
@arossert 
> I added the return_exceptions because I think that it will make the code look cleaner...

I guessed it, I did the same some time back 😄 
`return_exceptions=True` gives an impression of processing the exception object later, which we are not doing here. 

> The raise_for_status is needed since the API call can return with 401 status and an empty body, I assume that response.json will also fail in this case ...

Yeah, checking the json response automatically covers all the scenarios.

> Another option is to replace the try block with contextlib.suppress(BaseException)

I was not aware of that, looks cool. You can implement it.

> BTW, any reason you are using BaseException and not Exception?

That's how it was from the [beginning](https://github.com/dgzlopes/cloud-detect/commit/1611e77ee92f17d92f0b07e16636e32a83763343), I didn't give much thought either when I introduced `asyncio`.
The only advantage I see is that it stops nicely without any traceback and returns `False` when cancelled by user or another process. The downside is false positive when interrupted.

Reply:
Thanks for the information, it seems a bit weird to me since `BaseExceptoin` is rarely used.
Anyway, I have replaced the `try` with `contextlib`, if you are happy with this change I would appreciate it if you would merge and release a new version 🙏